HPX fails to configure with -DHPX_WITH_TESTS=OFF

## Expected Behavior

When configuring HPX for Phylanx builds, I disable the HPX tests with the ```-DHPX_WITH_TESTS=OFF``` flag.  That causes cmake to fail.

## Actual Behavior

CMake fails with several errors like:

```
CMake Error at cmake/HPX_AddPseudoDependencies.cmake:28 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "tests.examples".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
Call Stack (most recent call first):
  libs/cache/examples/CMakeLists.txt:8 (add_hpx_pseudo_dependencies)
```

## Steps to Reproduce the Problem

  1.  cmake ... -DHPX_WITH_TESTS=OFF ...

## Specifications

... Please describe your environment

  - HPX Version: master, "stable" tag
  - Platform (compiler, OS): linux, gcc 7.1

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -323,6 +323,16 @@ hpx_option(HPX_WITH_FAIL_COMPILE_TESTS BOOL
   "Create build system support for fail compile HPX tests (default ON)"
   ON CATEGORY "Build Targets")
 
+  # disable all tests if HPX_WITH_TESTS=OFF
+if(NOT HPX_WITH_TESTS)
+  hpx_set_option(HPX_WITH_TESTS_BENCHMARKS VALUE OFF FORCE)
+  hpx_set_option(HPX_WITH_TESTS_REGRESSIONS VALUE OFF FORCE)
+  hpx_set_option(HPX_WITH_TESTS_UNIT VALUE OFF FORCE)
+  hpx_set_option(HPX_WITH_TESTS_HEADERS VALUE OFF FORCE)
+  hpx_set_option(HPX_WITH_TESTS_EXTERNAL_BUILD VALUE OFF FORCE)
+  hpx_set_option(HPX_WITH_TESTS_EXAMPLES VALUE OFF FORCE)
+endif()
+
 # Enable IO-counters on linux systems only
 set(HPX_WITH_IO_COUNTERS_DEFAULT OFF)
 if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")

--- a/cmake/HPX_Option.cmake
+++ b/cmake/HPX_Option.cmake
@@ -48,3 +48,39 @@ function(hpx_option option type description default)
   set(${option}Category ${_category} CACHE INTERNAL "")
 endfunction()
 
+# simplify setting an option in cache
+function(hpx_set_option option)
+  set(options FORCE)
+  set(one_value_args VALUE TYPE HELPSTRING)
+  set(multi_value_args)
+  cmake_parse_arguments(HPX_SET_OPTION "${options}" "${one_value_args}" "${multi_value_args}" ${ARGN})
+
+  if(NOT DEFINED ${option})
+    hpx_error("attempting to set an undefined option: ${option}")
+  endif()
+
+  set(${option}_force)
+  if(HPX_SET_OPTION_FORCE)
+    set(${option}_force FORCE)
+  endif()
+
+  if(HPX_SET_OPTION_HELPSTRING)
+    set(${option}_description ${HPX_SET_OPTION_HELPSTRING})
+  else()
+    get_property(${option}_description CACHE "${option}" PROPERTY HELPSTRING)
+  endif()
+
+  if(HPX_SET_OPTION_TYPE)
+    set(${option}_type ${HPX_SET_OPTION_TYPE})
+  else()
+    get_property(${option}_type CACHE "${option}" PROPERTY TYPE)
+  endif()
+
+  if(DEFINED HPX_SET_OPTION_VALUE)
+    set(${option}_value ${HPX_SET_OPTION_VALUE})
+  else()
+    get_property(${option}_value CACHE "${option}" PROPERTY VALUE)
+  endif()
+
+  set(${option} ${${option}_value} CACHE ${${option}_type} "${${option}_description}" ${${option}_force})
+endfunction()

--- a/libs/CMakeLists.txt
+++ b/libs/CMakeLists.txt
@@ -5,6 +5,8 @@
 
 # This file is auto generated. Please do not edit manually.
 
+include(HPX_Message)
+
 set(HPX_LIBS
   cache
   config
@@ -12,6 +14,9 @@ set(HPX_LIBS
   CACHE INTERNAL "" FORCE
 )
 
+hpx_info("Configuring modules:")
+
 foreach(lib ${HPX_LIBS})
   add_subdirectory(${lib})
 endforeach()
+

--- a/libs/cache/CMakeLists.txt
+++ b/libs/cache/CMakeLists.txt
@@ -10,12 +10,15 @@ project(HPX.cache CXX)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
 include(HPX_AddDefinitions)
+include(HPX_Message)
 include(HPX_Option)
+
+hpx_info("  cache")
 
 hpx_option(HPX_CACHE_WITH_TESTS
   BOOL
-  "Build HPX cache module tests. (default: ON)"
-  ON ADVANCED
+  "Build HPX cache module tests. (default: ${HPX_WITH_TESTS})"
+  ${HPX_WITH_TESTS} ADVANCED
   CATEGORY "Modules")
 
 hpx_option(HPX_CACHE_WITH_DEPRECATION_WARNINGS
@@ -42,8 +45,6 @@ if(HPX_CACHE_WITH_COMPATIBILITY_HEADERS)
       DEFINE HPX_CACHE_HAVE_COMPATIBILITY_HEADERS
       NAMESPACE CACHE)
 endif()
-
-message(STATUS "cache: Configuring")
 
 set(headers
   ${CMAKE_CURRENT_SOURCE_DIR}/include/hpx/cache/local_cache.hpp
@@ -151,4 +152,3 @@ write_config_defines_file(
 add_subdirectory(examples)
 add_subdirectory(tests)
 
-message(STATUS "cache: Configuring done")

--- a/libs/cache/examples/CMakeLists.txt
+++ b/libs/cache/examples/CMakeLists.txt
@@ -3,7 +3,7 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-if (HPX_WITH_TESTS_EXAMPLES)
+if (HPX_WITH_TESTS AND HPX_WITH_TESTS_EXAMPLES)
   add_hpx_pseudo_target(tests.examples.cache)
   add_hpx_pseudo_dependencies(tests.examples tests.examples.cache)
 endif()

--- a/libs/cache/tests/CMakeLists.txt
+++ b/libs/cache/tests/CMakeLists.txt
@@ -3,7 +3,10 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-if (NOT HPX_WITH_TESTS)
+include(HPX_Option)
+
+if (NOT HPX_WITH_TESTS AND HPX_TOP_LEVEL)
+  hpx_set_option(HPX_CACHE_WITH_TESTS VALUE OFF FORCE)
   return()
 endif()
 if (NOT HPX_CACHE_WITH_TESTS)

--- a/libs/config/CMakeLists.txt
+++ b/libs/config/CMakeLists.txt
@@ -10,15 +10,16 @@ project(HPX.config CXX)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
 include(HPX_AddDefinitions)
+include(HPX_Message)
 include(HPX_Option)
+
+hpx_info("  config")
 
 hpx_option(HPX_CONFIG_WITH_TESTS
   BOOL
-  "Build HPX config module tests. (default: ON)"
-  ON ADVANCED
+  "Build HPX config module tests. (default: ${HPX_WITH_TESTS})"
+  ${HPX_WITH_TESTS} ADVANCED
   CATEGORY "Modules")
-
-message(STATUS "config: Configuring")
 
 set(headers
   ${CMAKE_CURRENT_SOURCE_DIR}/include/hpx/config.hpp
@@ -90,4 +91,3 @@ install(
 add_subdirectory(examples)
 add_subdirectory(tests)
 
-message(STATUS "config: Configuring done")

--- a/libs/config/examples/CMakeLists.txt
+++ b/libs/config/examples/CMakeLists.txt
@@ -3,6 +3,8 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-add_hpx_pseudo_target(tests.examples.config)
-add_hpx_pseudo_dependencies(tests.examples tests.examples.config)
+if (HPX_WITH_TESTS AND HPX_WITH_TESTS_EXAMPLES)
+  add_hpx_pseudo_target(tests.examples.config)
+  add_hpx_pseudo_dependencies(tests.examples tests.examples.config)
+endif()
 

--- a/libs/config/tests/CMakeLists.txt
+++ b/libs/config/tests/CMakeLists.txt
@@ -3,7 +3,10 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
+include(HPX_Option)
+
 if (NOT HPX_WITH_TESTS AND HPX_TOP_LEVEL)
+  hpx_set_option(HPX_CONFIG_WITH_TESTS VALUE OFF FORCE)
   return()
 endif()
 if (NOT HPX_CONFIG_WITH_TESTS)

--- a/libs/create_library_skeleton.py
+++ b/libs/create_library_skeleton.py
@@ -68,25 +68,24 @@ project(HPX.{lib_name} CXX)
 list(APPEND CMAKE_MODULE_PATH "${{CMAKE_CURRENT_SOURCE_DIR}}/cmake")
 
 include(HPX_AddDefinitions)
+include(HPX_Message)
 include(HPX_Option)
+
+hpx_info("  {lib_name}")
 
 hpx_option(HPX_{lib_name_upper}_WITH_TESTS
   BOOL
-  "Build HPX {lib_name} module tests. (default: ON)"
-  ON ADVANCED
+  "Build HPX {lib_name} module tests. (default: ${HPX_WITH_TESTS})"
+  ${HPX_WITH_TESTS} ADVANCED
   CATEGORY "Modules")
-
-message(STATUS "{lib_name}: Configuring")
 
 add_subdirectory(examples)
 add_subdirectory(src)
 add_subdirectory(tests)
-
-message(STATUS "{lib_name}: Configuring done")
 '''
 
 examples_cmakelists_template = cmake_header + f'''
-if (HPX_WITH_TESTS_EXAMPLES)
+if (HPX_WITH_TESTS AND HPX_WITH_TESTS_EXAMPLES)
   add_hpx_pseudo_target(tests.examples.{lib_name})
   add_hpx_pseudo_dependencies(tests.examples tests.examples.{lib_name})
 endif()
@@ -94,7 +93,10 @@ endif()
 '''
 
 tests_cmakelists_template = cmake_header + f'''
-if (NOT HPX_WITH_TESTS)
+include(HPX_Option)
+
+if (NOT HPX_WITH_TESTS AND HPX_TOP_LEVEL)
+  hpx_set_option(HPX_{lib_name_upper}_WITH_TESTS VALUE OFF FORCE)
   return()
 endif()
 if (NOT HPX_{lib_name_upper}_WITH_TESTS)
@@ -205,6 +207,8 @@ libs_cmakelists = cmake_header + f'''
 '''
 
 libs_cmakelists += '''
+include(HPX_Message)
+
 set(HPX_LIBS
 '''
 for lib in libs:
@@ -213,6 +217,8 @@ for lib in libs:
 libs_cmakelists += '  CACHE INTERNAL "" FORCE\n)\n'
 
 libs_cmakelists += '''
+hpx_info("Configuring modules:")
+
 foreach(lib ${HPX_LIBS})
   add_subdirectory(${lib})
 endforeach()

--- a/libs/preprocessor/CMakeLists.txt
+++ b/libs/preprocessor/CMakeLists.txt
@@ -10,12 +10,15 @@ project(HPX.preprocessor CXX)
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
 
 include(HPX_AddDefinitions)
+include(HPX_Message)
 include(HPX_Option)
+
+hpx_info("  preprocessor")
 
 hpx_option(HPX_PREPROCESSOR_WITH_TESTS
   BOOL
-  "Build HPX preprocessor module tests. (default: ON)"
-  ON ADVANCED
+  "Build HPX preprocessor module tests. (default: ${HPX_WITH_TESTS})"
+  ${HPX_WITH_TESTS} ADVANCED
   CATEGORY "Modules")
 
 hpx_option(HPX_PREPROCESSOR_WITH_DEPRECATION_WARNINGS
@@ -41,8 +44,6 @@ if(HPX_PREPROCESSOR_WITH_COMPATIBILITY_HEADERS)
     DEFINE HPX_PREPROCESSOR_HAVE_COMPATIBILITY_HEADERS
     NAMESPACE PREPROCESSOR)
 endif()
-
-message(STATUS "preprocessor: Configuring")
 
 set(headers
   ${CMAKE_CURRENT_SOURCE_DIR}/include/hpx/preprocessor.hpp
@@ -130,4 +131,3 @@ write_config_defines_file(
   NAMESPACE PREPROCESSOR
   FILENAME "${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/hpx/preprocessor/config/defines.hpp")
 
-message(STATUS "preprocessor: Configuring done")

--- a/libs/preprocessor/tests/CMakeLists.txt
+++ b/libs/preprocessor/tests/CMakeLists.txt
@@ -3,7 +3,8 @@
 # Distributed under the Boost Software License, Version 1.0. (See accompanying
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-if (NOT HPX_WITH_TESTS)
+if (NOT HPX_WITH_TESTS AND HPX_TOP_LEVEL)
+  hpx_set_option(HPX_PREPROCESSOR_WITH_TESTS VALUE OFF FORCE)
   return()
 endif()
 if (NOT HPX_PREPROCESSOR_WITH_TESTS)

